### ISSUE TITLE
Handle multipart image upload in a function

### ISSUE THREAD

**Issue by toxsick** [redwoodjs/redwood#1543]
Hey guys,

first of all thanks for redwood, it is a delight!

I spend a few hours trying to get a multipart image upload function working, but the uploaded images are always corrupted. First I tried to use multer (with express and serverless), but because [it seems to have problems](https://github.com/expressjs/multer/issues/770) with lambda I tried other libraries like `lambda-multipart-parser`:

```js
import fs from 'fs';
import multipartParser from 'lambda-multipart-parser';

export const handler = async (event, context) => {
  const { files } = await multipartParser.parse(event);
  const file = files[0];
  await fs.writeFileSync(`./${file.filename}`, file.content);
  return {
    statusCode: 200,
  };
};
```
Sadly no success, the **uploaded files are always broken**.  I'm not sure I can tell u more right now then they are broken on the other side :)

Further I am only trying this in development mode. Could it happen that a function that works locally ([redwood uses express for dev I think](https://github.com/redwoodjs/redwood/blob/main/packages/api-server/src/http.ts)) and it is broken when I deploy it to netlify / lambda?

Sorry guys, I need a little bit of directions here and the serverless stuff is still new to me.

regards and thx

**Comment by peterp** [comment]
Hey @toxsick, 

> Further I am only trying this in development mode. Could it happen that a function that works locally (redwood uses express for dev I think) and it is broken when I deploy it to netlify / lambda?

That could be the case - we're emulating aws' lambdas and it could be that this is something I didn't make provision for yet.

**Comment by peterp** [comment]
My gut feeling is that we don't have all the correct things in `event` to allow that multipart library to work.

**Comment by toxsick** [comment]
Mhhh alright, my previous approach with multer was something like this, but the resulting file was even more broken (file size didn't match):

```js
import express from 'express';
import serverless from 'serverless-http';
import multer from 'multer';
import multerMinio from 'multer-minio-storage-engine';
import { s3 } from '../lib/s3';

const app = express();

const upload = multer({
  storage: multerMinio({
    minio: s3,
    bucketName: 'images',
    metaData(req, file, cb) {
      cb(null, { mimetype: file.mimetype });
    },
    objectName(req, file, cb) {
      const newFileName = `${new Date().toISOString()}-${file.originalname}`;
      cb(null, newFileName);
    },
  }),
});

app.post('/upload', upload.array('images', 1), (req, res) => {
  res.json({
    message: 'File uploaded succesfully.',
  });
});

module.exports.handler = serverless(app);
```

@peterp because there seem to be some problems around multipart upload and lambda this could be the case here too. One thing that comes up is that for AWS users have to enable "Lambda Proxy Integration" ([some issue](https://github.com/francismeynard/lambda-multipart-parser/issues/3), [docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html)), but not sure what that does?! Perhaps it helps for context.

**Comment by toxsick** [comment]
just for reference: https://community.netlify.com/t/functions-issues-parsing-images-from-multipart-form-data/3068/23

**Comment by peterp** [comment]
@toxsick Ah, seems like trying to do it like this this is going to be a slippery slope, why not upload to s3 directly?

Here's an example of how this works in next.js, it would be almost identical via RedwoodJS: https://github.com/leerob/nextjs-aws-s3/blob/main/pages/api/upload-url.js

**Comment by Tobbe** [comment]
There has been some talk about S3 image uploads in the forums. https://community.redwoodjs.com/t/aws-s3-file-uploads/1518

TLDR; Use ImageKit instead ^^
(Or go with @dac09's suggestion of signed URLs)

**Comment by toxsick** [comment]
@peterp That is exactly what I am doing right now :) I can post my function code here when I am done for reference.

@Tobbe  Thanks for the info. I got it working with Filepond and direct upload (with presigend urls) to S3 / Minio also.

**Comment by toxsick** [comment]
Ok here as promised my implementation with [filepond](https://pqina.nl/filepond/) and [minio](https://min.io/). In the assets function I'm using `serverless` and `express`, which is not really necessary, but for now it does the job:

```js
// assets.ts

import express from 'express';
import serverless from 'serverless-http';
import { s3 } from '../lib/s3';

const app = express();
const router = express.Router();

app.use('/assets', router);

router.get('/upload-url', async (req, res) => {
  const { assetId, assetFileExtension, assetType } = req.query;

  // TODO: handle invalid query

  const uploadUrl = await s3.presignedPutObject(
    'images',
    `${assetType}/${assetId}.${assetFileExtension}`,
    // TODO: set expiry
  );
  res.json({ uploadUrl });
});

module.exports.handler = serverless(app);

```

``` js
// UploadImage.tsx

import { FilePond, registerPlugin } from 'react-filepond';

import FilePondPluginImageExifOrientation from 'filepond-plugin-image-exif-orientation';
import FilePondPluginImagePreview from 'filepond-plugin-image-preview';

import FilePondPluginFileValidateType from 'filepond-plugin-file-validate-type';
import FilePondPluginImageCrop from 'filepond-plugin-image-crop';
import FilePondPluginImageResize from 'filepond-plugin-image-resize';
import FilePondPluginImageTransform from 'filepond-plugin-image-transform';

import 'filepond/dist/filepond.min.css';
import 'filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css';

// Register the plugins
registerPlugin(
  // validates files based on input type
  FilePondPluginFileValidateType,
  // corrects mobile image orientation
  FilePondPluginImageExifOrientation,
  // previews the image
  FilePondPluginImagePreview,
  // crops the image to a certain aspect ratio
  FilePondPluginImageCrop,
  // resizes the image to fit a certain size
  FilePondPluginImageResize,
  // applies crop and resize information on the client
  FilePondPluginImageTransform,
);

type AssetType = 'author' | 'post';

type UploadImageProps = {
  assetId: string;
  assetType: AssetType;
  className?: string;
  existingAssetUrl?: string;
  imageResizeTargetHeight?: number;
  imageResizeTargetWidth?: number;
};

const UploadImage: React.FunctionComponent<UploadImageProps> = ({
  assetId,
  assetType,
  className,
  existingAssetUrl,
  imageResizeTargetHeight,
  imageResizeTargetWidth,
}: UploadImageProps) => {
  // eslint-disable-next-line no-underscore-dangle
  const serverUrl = `${window.__REDWOOD__API_PROXY_PATH}/assets`;

  const getUploadUrl = async (file) => {
    const assetFileExtension = file.name.split('.').pop();
    const res = await fetch(
      `${serverUrl}/upload-url?assetId=${assetId}&assetFileExtension=${assetFileExtension}&assetType=${assetType}`,
    );
    const { uploadUrl } = await res.json();
    return uploadUrl;
  };

  const processFile = async (
    _fieldName,
    file,
    _metadata,
    load,
    error,
    // progress,
    // abort,
  ) => {
    try {
      // TODO: all error handling + progress
      const uploadUrl = await getUploadUrl(file);
      await fetch(uploadUrl, {
        method: 'PUT',
        body: file,
      });
      load(assetId);
    } catch (err) {
      error(err);
    }
  };

  const loadFile = async (source, load, error) => {
    try {
      const res = await fetch(existingAssetUrl);
      const blob = await res.blob();
      // see: https://github.com/pqina/filepond/blob/master/src/js/utils/getFileFromBlob.js
      blob.name = assetId;
      load(blob);
    } catch (err) {
      error(err);
    }
  };

  const initalFiles = existingAssetUrl
    ? // this tells filepond that an image exists that
      // has to be fetched from the server using the
      // load function
      // see: Setting initial files
      // https://pqina.nl/filepond/docs/patterns/api/filepond-object/#creating-a-filepond-instance
      [
        {
          source: assetId,
          options: {
            type: 'local',
          },
        },
      ]
    : [];

  return (
    // @ts-expect-error filepond has ts errors here for now
    <FilePond
      acceptedFileTypes={['image/jpeg', 'image/png', 'image/webp']}
      className={className}
      files={initalFiles}
      imageCropAspectRatio="1:1"
      imageResizeTargetHeight={imageResizeTargetHeight}
      imageResizeTargetWidth={imageResizeTargetWidth}
      // do not upscale images when small than target size
      imageResizeUpscale={false}
      name="asset"
      styleButtonProcessItemPosition="center bottom"
      styleButtonRemoveItemPosition="center bottom"
      styleLoadIndicatorPosition="center bottom"
      stylePanelLayout="compact circle"
      styleProgressIndicatorPosition="center bottom"
      server={{
        url: serverUrl,
        process: processFile,
        load: loadFile,
        revert: null,
        restore: null,
        fetch: null,
      }}
    />
  );
};

UploadImage.defaultProps = {
  className: null,
  existingAssetUrl: null,
  imageResizeTargetHeight: 1000,
  imageResizeTargetWidth: 1000,
};

export default UploadImage;

```

Hope it helps somebody who ends up here. For me you can consider this issue resolved.

Thanks for the support guys!
